### PR TITLE
Reenables scissor on shadow for individual clears

### PR DESCRIPTION
**Pull request by Kuranes**
test/benchmarker on samsung s6

### CODE REVIEW

**Comment by Kuranes** [comment]
@cedricpinson to check the scissor thing.

**Comment by cedricpinson** [comment]
I feel that you put back the code of scissor when we tried a few weeks ago, but in the meantime this commit https://github.com/cedricpinson/osgjs/pull/869/commits/d27e3e2c02ba954eed44d9d307dd567d880bca0a added scissor in camera to handle better scissor behavior with RTT
I added extra comment in the code

File: sources/osg/RenderStage.js [hunk 1]
```diff
@@ -282,8 +282,16 @@ utils.createPrototypeObject(
             }
             state.applyAttribute(this._viewport);
 
-            if (this._scissor) {
-                state.applyAttribute(this._scissor);
+            // fragment clipping
+            var scissor;
```
Comment (cedricpinson) [inline sources/osg/RenderStage.js]: This change is useless if you use the setScissor method in camera

File: sources/osgShadow/ShadowMapAtlas.js [hunk 2]
```diff
@@ -20,6 +22,7 @@ var ShadowMapAtlas = function(settings) {
     this._lights = [];
     this._shadowMaps = [];
     this._viewportDimension = [];
+    this._scissors = [];
```
Comment (cedricpinson) [inline sources/osgShadow/ShadowMapAtlas.js]: you probably dont need this variable anymore maybe you can just reference the camera

File: sources/osgShadow/ShadowMapAtlas.js [hunk 3]
```diff
@@ -293,21 +296,21 @@ utils.createPrototypeObject(
 
                 if (this._viewportDimension.length <= i) {
                     this._viewportDimension.push(vec4.fromValues(x, y, mapSizeX, mapSizeY));
+
+                    var st = shadowMap.getCamera().getOrCreateStateSet();
+                    var scissor = new Scissor(x, y, mapSizeX, mapSizeY);
+                    this._scissors.push(scissor);
```
Comment (cedricpinson) [inline sources/osgShadow/ShadowMapAtlas.js]: should be replaced by shadowMap.getCamera().setScissor(scissor);

File: sources/osgShadow/ShadowMapAtlas.js [hunk 4]
```diff
@@ -293,21 +296,21 @@ utils.createPrototypeObject(
...
+                    st.setAttributeAndModes(scissor, StateAttribute.ON | StateAttribute.OVERRIDE);
```
Comment (cedricpinson) [inline sources/osgShadow/ShadowMapAtlas.js]: why did you need the ON|OVERRIDE ?

File: sources/osgShadow/ShadowMapAtlas.js [hunk 5]
```diff
@@ -293,21 +296,21 @@ utils.createPrototypeObject(
...
                 } else {
                     vec4.set(this._viewportDimension[i], x, y, mapSizeX, mapSizeY);
+                    this._scissors[i].setScissor(x, y, mapSizeX, mapSizeY);
```
Comment (cedricpinson) [inline sources/osgShadow/ShadowMapAtlas.js]: Could be changed by the camera.getScissor().setScissor(x,y,mapSizeX,mapSizeY);

File: sources/osgShadow/ShadowMapAtlas.js [hunk 6]
```diff
@@ -409,6 +409,13 @@ utils.createPrototypeObject(
                     idx = this._lights.indexOf(light);
                     if (idx !== -1) this._lights.splice(idx, 1);
 
+                    if (this._viewportDimension.length > idx) {
+                        this._viewportDimension.splice(idx, 1);
```
Comment (cedricpinson) [inline sources/osgShadow/ShadowMapAtlas.js]: why not just keeping a camera array ? inside you can have viewport and scissor

Comment (Kuranes) [reply sources/osgShadow/ShadowMapAtlas.js]: it's vec4 I store and keep. 
(it's the data of the uniform "mapSize" for shadow receiving (clamping on tiles in atlas))
